File: src/gdt.c
/* GDT: flat 64-bit kernel code/data segments.
 * Limine already loads its own GDT; we install ours so segment layout is
 * stable for user segments (M7) and TSS (M6).
 *
 * Layout:
 *   0x00 null
 *   0x08 kernel code (64-bit, DPL 0)
 *   0x10 kernel data          (DPL 0)
 *   0x18 user data            (DPL 3)   -- reserved, used in M7
 *   0x20 user code (64-bit)   (DPL 3)   -- reserved, used in M7
 *   0x28 TSS (16 bytes)        -- ring3 -> ring0 interrupt stack
 */
#include "gdt.h"
#include "kprintf.h"
#include "lib/string.h"
#include <stdint.h>

static uint64_t gdt[7];

static const uint16_t KERNEL_CS = 0x08;
static const uint16_t KERNEL_DS = 0x10;

struct tss {
    uint32_t reserved0;
    uint64_t rsp0, rsp1, rsp2;
    uint64_t reserved1;
    uint64_t ist1, ist2, ist3, ist4, ist5, ist6, ist7;
    uint64_t reserved2;
    uint16_t reserved3;
    uint16_t iomap_base;
} __attribute__((packed));

static struct tss tss;
static uint64_t boot_irq_stack[2048];   /* 16KiB fallback for rsp0 */

struct gdtr {
    uint16_t limit;
    uint64_t base;
} __attribute__((packed));

void gdt_set_kernel_stack(uint64_t rsp0) {
    tss.rsp0 = rsp0;
}

static uint64_t make_tss_desc_low(uint64_t base, uint32_t limit) {
    return ((uint64_t)(limit & 0xFFFF))
         | ((base & 0xFFFFFF) << 16)
         | (0x89ULL << 40)                       /* present, type=64-bit TSS */
         | (((uint64_t)(limit >> 16) & 0xF) << 48)
         | (((base >> 24) & 0xFF) << 56);
}

void gdt_init(void) {
    gdt[0] = 0x0000000000000000ULL; /* null          */
    gdt[1] = 0x00AF9A000000FFFFULL; /* kernel code   */
    gdt[2] = 0x00AF92000000FFFFULL; /* kernel data   */
    gdt[3] = 0x00AFF2000000FFFFULL; /* user data     */
    gdt[4] = 0x00AFFA000000FFFFULL; /* user code     */

    tss.rsp0       = (uint64_t)&boot_irq_stack + sizeof boot_irq_stack;
    tss.iomap_base = sizeof tss;
    uint64_t base  = (uint64_t)&tss;
    uint32_t limit = sizeof(tss) - 1;
    gdt[5] = make_tss_desc_low(base, limit);
    gdt[6] = base >> 32;

    struct gdtr gdtr = {
        .limit = sizeof(gdt) - 1,
        .base  = (uint64_t)&gdt,
    };

    __asm__ volatile (
        "lgdt %0\n\t"
        /* reload data segments */
        "mov %1, %%ds\n\t"
        "mov %1, %%es\n\t"
        "mov %1, %%fs\n\t"
        "mov %1, %%gs\n\t"
        "mov %1, %%ss\n\t"
        /* reload cs via far return */
        "pushq %2\n\t"
        "leaq 1f(%%rip), %%rax\n\t"
        "pushq %%rax\n\t"
        "lretq\n\t"
        "1:\n\t"
        :
        : "m"(gdtr), "r"(KERNEL_DS), "r"((uint64_t)KERNEL_CS)
        : "rax", "memory");

    KLOG_INFO("gdt: installed (cs=%#lx ds=%#lx)",
              (unsigned long)KERNEL_CS, (unsigned long)KERNEL_DS);

    __asm__ volatile ("ltr %w0" : : "r"((uint16_t)0x28));
    KLOG_INFO("gdt: tss loaded, rsp0=%p", (void *)tss.rsp0);
}

File: kernel-rs/src/gdt.rs
//! GDT with kernel/user segments and a TSS for ring3 -> ring0 stack switches.
//!
//! Layout (matches the old C implementation):
//!   0x00 null
//!   0x08 kernel code (64-bit, DPL 0)
//!   0x10 kernel data          (DPL 0)
//!   0x18 user data            (DPL 3)
//!   0x20 user code (64-bit)   (DPL 3)
//!   0x28 TSS (16 bytes)

use core::arch::asm;
use core::cell::UnsafeCell;
use core::ptr::{addr_of, addr_of_mut};

pub const KERNEL_CS: u16 = 0x08;
pub const KERNEL_DS: u16 = 0x10;
pub const USER_DS: u16 = 0x18;
pub const USER_CS: u16 = 0x20;
pub const TSS_SEL: u16 = 0x28;

#[repr(C, packed)]
pub struct Tss {
    reserved0: u32,
    pub rsp0: u64,
    rsp1: u64,
    rsp2: u64,
    reserved1: u64,
    ist: [u64; 7],
    reserved2: u64,
    reserved3: u16,
    iomap_base: u16,
}

struct TssCell(UnsafeCell<Tss>);
unsafe impl Sync for TssCell {}

struct GdtCell(UnsafeCell<[u64; 7]>);
unsafe impl Sync for GdtCell {}

static GDT: GdtCell = GdtCell(UnsafeCell::new([0; 7]));
static TSS: TssCell = TssCell(UnsafeCell::new(Tss {
    reserved0: 0,
    rsp0: 0,
    rsp1: 0,
    rsp2: 0,
    reserved1: 0,
    ist: [0; 7],
    reserved2: 0,
    reserved3: 0,
    iomap_base: 0,
}));

static mut BOOT_IRQ_STACK: [u8; 16384] = [0; 16384]; // 16KiB fallback for rsp0

#[repr(C, packed)]
struct Gdtr {
    limit: u16,
    base: u64,
}

pub fn set_kernel_stack(rsp0: u64) {
    unsafe { (*TSS.0.get()).rsp0 = rsp0 };
}

fn make_tss_desc_low(base: u64, limit: u32) -> u64 {
    (limit & 0xFFFF) as u64
        | ((base & 0xFF_FFFF) << 16)
        | (0x89u64 << 40) // present, type=64-bit TSS
        | (((limit >> 16) as u64 & 0xF) << 48)
        | ((base >> 24 & 0xFF) << 56)
}

pub fn init() {
    unsafe {
        let gdt = &mut *GDT.0.get();
        let tss = &mut *TSS.0.get();

        gdt[0] = 0x0000_0000_0000_0000; // null
        gdt[1] = 0x00AF_9A00_0000_FFFF; // kernel code
        gdt[2] = 0x00AF_9200_0000_FFFF; // kernel data
        gdt[3] = 0x00AF_F200_0000_FFFF; // user data
        gdt[4] = 0x00AF_FA00_0000_FFFF; // user code

        tss.rsp0 = addr_of_mut!(BOOT_IRQ_STACK) as u64 + BOOT_IRQ_STACK.len() as u64;
        tss.iomap_base = core::mem::size_of::<Tss>() as u16;
        let base = tss as *mut Tss as u64;
        let limit = (core::mem::size_of::<Tss>() - 1) as u32;
        gdt[5] = make_tss_desc_low(base, limit);
        gdt[6] = base >> 32;

        let gdtr = Gdtr {
            limit: (core::mem::size_of::<[u64; 7]>() - 1) as u16,
            base: gdt.as_ptr() as u64,
        };

        asm!(
            "lgdt [{gdtr}]",
            "mov ds, {ds:x}",
            "mov es, {ds:x}",
            "mov fs, {ds:x}",
            "mov gs, {ds:x}",
            "mov ss, {ds:x}",
            "push {cs}",
            "lea rax, [rip + 2f]",
            "push rax",
            "retfq",
            "2:",
            gdtr = in(reg) &gdtr,
            ds = in(reg) KERNEL_DS as u64,
            cs = in(reg) KERNEL_CS as u64,
            out("rax") _,
        );

        crate::klog_info!("gdt: installed (cs={:#x} ds={:#x})", KERNEL_CS, KERNEL_DS);

        asm!("ltr {sel:x}", sel = in(reg) TSS_SEL, options(nomem, nostack));
        let rsp0 = addr_of!((*TSS.0.get()).rsp0).read_unaligned();
        crate::klog_info!("gdt: tss loaded, rsp0={:#x}", rsp0);
    }
}
